File: 814_prune_tree/prune_tree.go
/*
 * @Date: 2022-07-21
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-07-21
 * @FilePath: /algorithm/814_prune_tree/prune_tree.go
 */

package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pruneTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	root.Left = pruneTree(root.Left)
	root.Right = pruneTree(root.Right)
	if root.Left == nil && root.Right == nil && root.Val == 0 {
		return nil
	}
	return root
}

func main() {
	pruneTree(&TreeNode{Val: 1, Left: nil, Right: &TreeNode{Val: 0, Left: &TreeNode{Val: 0}, Right: &TreeNode{Val: 1}}})
	pruneTree(&TreeNode{Val: 1, Left: &TreeNode{Val: 0, Left: &TreeNode{Val: 0}, Right: &TreeNode{Val: 0}}, Right: &TreeNode{Val: 0, Left: &TreeNode{Val: 0}, Right: &TreeNode{Val: 1}}})
	pruneTree(&TreeNode{Val: 1, Left: &TreeNode{Val: 1, Left: &TreeNode{Val: 1, Left: &TreeNode{Val: 1}}, Right: &TreeNode{Val: 1}}, Right: &TreeNode{Val: 0, Left: &TreeNode{Val: 0}, Right: &TreeNode{Val: 1}}})
}

File: cpp/814_prune_tree/prune_tree.cpp
/*
 * @Date: 2022-07-21
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-07-21
 * @FilePath: /algorithm/814_prune_tree/prune_tree.cpp
 */

#include <cassert>
#include <iostream>

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode *pruneTree(TreeNode *root) {
    if (!root) {
      return nullptr;
    }
    root->left = pruneTree(root->left);
    root->right = pruneTree(root->right);
    if (!root->left && !root->right && !root->val) {
      return nullptr;
    }
    return root;
  }
};

int main() {
  Solution().pruneTree(new TreeNode(1, nullptr, new TreeNode(0, new TreeNode(0), new TreeNode(0))));
  Solution().pruneTree(new TreeNode(1, new TreeNode(0, new TreeNode(0), new TreeNode(0)),
                                    new TreeNode(1, new TreeNode(0), new TreeNode(1))));
  Solution().pruneTree(new TreeNode(1, new TreeNode(1, new TreeNode(1, new TreeNode(0), nullptr), new TreeNode(1)),
                                    new TreeNode(0, new TreeNode(1), new TreeNode(0))));
}
